src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -354,15 +355,34 @@ CreateDistributedTable(Oid relationId, Var *distributionColumn, char distributio
 	 * given relation is involved, then we undistribute the relation and finally
 	 * we re-create dropped foreign keys at the end of this function.
 	 */
-	List *fKeyCreationCommandsRelationInvolved = NIL;
+	List *droppedFKeyCreationCommands = NIL;
 	if (IsCitusTableType(relationId, CITUS_LOCAL_TABLE))
 	{
 		/* store foreign key creation commands that relation is involved */
-		fKeyCreationCommandsRelationInvolved =
-			GetFKeyCreationCommandsRelationInvolved(relationId);
+		droppedFKeyCreationCommands =
+			GetFKeyCreationCommandsRelationInvolvedWithTableType(relationId,
+																 INCLUDE_ALL_TABLE_TYPES);
 		relationId = DropFKeysAndUndistributeTable(relationId);
 	}
 
+	/*
+	 * To support foreign keys between reference tables and local tables,
+	 * we drop & re-define foreign keys at the end of this function so
+	 * that ALTER TABLE hook does the necessary job, which means converting
+	 * local tables to citus local tables to properly support such foreign
+	 * keys.
+	 */
+	else if (distributionMethod == DISTRIBUTE_BY_NONE &&
```

Comment:
maybe IsCitusTableType(relationId, CITUS_REFERENCE_TABLE) just for clarity

Reply:
we are actually just creating the reference table, so we don't have this info in metadata cache :/